src/github.com/getlantern/peerscanner/host.go

```diff
@@ -0,0 +1,407 @@
+	if err != nil {
+		return fmt.Errorf("Unable to register rotations: %v", err)
+	}
+	return nil
+}
+
+func (h *host) registerHost() error {
+	if h.record != nil {
+		log.Tracef("Host already registered, no need to re-register: %v", h)
+		return nil
+	}
+
+	log.Debugf("Registering %v", h)
+
+	rec, err := cfutil.Register(h.name, h.ip)
+	if err == nil || isDuplicateError(err) {
+		h.record = rec
+		err = nil
+	}
+	return err
+}
+
+func (h *host) registerToRotations() error {
+	for _, group := range h.groups {
+		err := group.register(h)
+		if err != nil && !isDuplicateError(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *host) deregister() {
+	h.deregisterHost()
+	h.deregisterFromRotations()
+}
+
+func (h *host) deregisterHost() {
+	if h.record == nil {
+		log.Debugf("Host not registered, no need to deregister: %v", h)
+		return
+	}
+
+	if h.isFallback() {
+		log.Debugf("Currently not deregistering fallbacks like %v", h)
+		return
+	}
+
+	log.Debugf("Deregistering %v", h)
+	h.doDeregisterHost()
+}
+
+func (h *host) doDeregisterHost() {
+	err := cfutil.DestroyRecord(h.record)
+	if err != nil {
+		log.Errorf("Unable to deregister host %v: %v", h, err)
+		return
+	}
+
+	h.record = nil
+}
+
+func (h *host) deregisterFromRotations() {
+	for _, group := range h.groups {
+		group.deregister(h)
+	}
+}
+
+func (h *host) fullName() string {
+	return h.name + ".getiantem.org"
+}
+
+func (h *host) isFallback() bool {
+	return isFallback(h.name)
+}
+
+func (h *host) isAbleToProxy() (bool, bool, error) {
+	// Check whether or not we can proxy a few times
+	var lastErr error
+	for i := 0; i < proxyAttempts; i++ {
+		success, connectionRefused, err := h.doIsAbleToProxy()
+		if err != nil {
+			log.Debugf("Error testing %v: %v", h, err.Error())
+		}
+		lastErr = err
+		if success || connectionRefused {
+			// If we've succeeded, or our connection was flat-out refused, don't
+			// bother trying to proxy again
+			return success, connectionRefused, lastErr
+		}
+	}
+	return false, false, lastErr
+}
+
+func (h *host) doIsAbleToProxy() (bool, bool, error) {
+	// First just try a plain TCP connection. This is useful because the underlying
+	// TCP-level error is consumed in the flashlight layer, and we need that
+	// to be accessible on the client side in the logic for deciding whether
+	// or not to display the port mapping message.
+	addr := h.ip + ":443"
+	conn, err := net.DialTimeout("tcp", addr, dialTimeout)
+	if err != nil {
+		err2 := fmt.Errorf("Unable to connect to %v: %v", addr, err)
+		return false, strings.Contains(err.Error(), "connection refused"), err2
+	}
+	conn.Close()
+
+	// Now actually try to proxy an http request
+	site := testSites[rand.Intn(len(testSites))]
+	resp, err := h.proxiedClient.Head("http://" + site)
+	if err != nil {
+		return false, false, fmt.Errorf("Unable to make proxied HEAD request to %v: %v", site, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 301 {
```

Comment:
>  My rule of thumb on this is that if it's used more than once, make it a constant, otherwise don't.

Absolutely! maybe we could look for good candidates to be promoted to constants later.